reports/helpers.py

```diff
@@ -0,0 +1,252 @@
+        end_of_month = invoice_month + timedelta(days=month_days - 1)
+
+    terminated_after_billing_month = appserver.terminated and _trunc_datetime(appserver.terminated) > invoice_month
+    if not appserver.terminated or terminated_after_billing_month:
+        last_billing_day = end_of_month
+    else:
+        last_billing_day = appserver.terminated
+
+    if appserver.created < invoice_month:
+        first_billing_day = invoice_month
+    else:
+        first_billing_day = appserver.created
+
+    return first_billing_day, last_billing_day
+
+
+def generate_charge_details(appserver, invoice_month):
+    """
+    Given a single AppServer, generate all the charges and their details
+    for a given month
+    :param appserver: An OpenEdxAppServer
+    :param invoice_month: The beginning of the month we want to issue usages for.
+    :return: A dictionary all charges details. Example:
+                {
+                    'name': 'AppServer 1',
+                    'billing_start': 1 - 9 - 2018,
+                    'billing_end': 30 - 9 - 2018,
+                    'days': 30,
+                    'charge': 300,
+                }
+    """
+    days = 0
+    charge = 0
+    billing_start, billing_end = get_billing_dates(appserver, invoice_month)
+
+    if billing_start and billing_end:
+        # Adding one to the days to include the first day
+        days = billing_end.day - billing_start.day + 1
+        charge = settings.BILLING_RATE * days
+
+    return {
+        'name': appserver.name,
+        'billing_start': billing_start,
+        'billing_end': billing_end,
+        'days': days,
+        'charge': charge,
+    }
+
+
+def get_instance_charges(instance, invoice_month):
+    """
+    Given an instance, this function will be responsible of generating
+    all the charges for the given instance's AppServers
+    :param instance: An openEdxInstance object
+    :param invoice_month: The beginning of the month we want to issue usages for.
+    :return: A pair of the AppServers total usage and a list of the AppServers
+             usages details. Example:
+                [
+                    {
+                        'name': 'AppServer 1',
+                        'billing_start': 1 - 9 - 2018,
+                        'billing_end': 30 - 9 - 2018,
+                        'days': 30,
+                        'charge': 300,
+                    },
+                    {
+                        'name': 'AppServer 2',
+                        'billing_start': 1 - 9 - 2018,
+                        'billing_end': 10 - 9 - 2018,
+                        'days': 10,
+                        'charge': 200,
+                    }
+                ]
+    """
+    appservers_charges = []
+    appservers_total = 0
+
+    if invoice_month > timezone.now():
+        return appservers_charges, appservers_total
+
+    for server in instance.appserver_set.all():
```

Comment:
What happens with servers that failed configuration? They will still be billed, right? They were created but were not „terminated“, so now they are accounted as active and billed.
They should be excluded. We don't need „active appservers“; I think it's best to explicitly get „terminated“ + „running“ appservers.
From this `reports` app, you can filter by AppServer status, in order to get only the `Running` and `Terminated` ones. You can do it in a QuerySet if it's clean, otherwise [here's how](https://github.com/open-craft/opencraft/blob/70235cd077eee7c44b6bb4785bd9a76dad4efd57/registration/approval.py#L66). I think we don't need to bill the other statuses (Error, ConfiguringServer, …) though we may rethink this in the future.

This will also require a test…